samples/boards/intel_adsp/core_power_toggle/src/main.c

```diff
@@ -0,0 +1,281 @@
+	/* TODO: before requesting core shut down check if it's not actively used */
+	if (!pm_state_force(id, &(struct pm_state_info){PM_STATE_SOFT_OFF, 0, 0})) {
+		LOG_ERR("failed to set PM_STATE_SOFT_OFF on core %d", id);
+		goto out;
+	}
+
+	/* Primary core will be turn off by the host after it enter SOFT_OFF state */
+	if (cpu_is_primary(id)) {
+		goto out;
+	}
+
+	/* Broadcasting interrupts to other cores. */
+	arch_sched_ipi();
+
+	uint64_t timeout = k_cycle_get_64() + k_ms_to_cyc_ceil64(10000);
+
+	/* Waiting for secondary core to enter idle state */
+	printk("waiting for core %d to enter idle, timeout %llu, cycle %llu\n", id, timeout,
+	       k_cycle_get_64());
+	while (arch_cpu_active(id) && (k_cycle_get_64() < timeout)) {
+		k_busy_wait(10000);
+	}
+	printk("done waiting for core %d, active? %d\n", id, arch_cpu_active(id));
+
+	if (arch_cpu_active(id)) {
+		printk("core %d did not enter idle state\n", id);
+		goto out;
+	}
+
+	if (soc_adsp_halt_cpu(id) != 0) {
+		printk("failed to disable core %d", id);
+	}
+#endif /* CONFIG_PM */
+out:
+	return;
+}
+
+int cpu_is_core_enabled(int id)
+{
+	return arch_cpu_active(id);
+}
+
+int cpu_enabled_cores(void)
+{
+	unsigned int i;
+	int mask = 0;
+
+	for (i = 0; i < CONFIG_MP_MAX_NUM_CPUS; i++) {
+		if (arch_cpu_active(i)) {
+			mask |= BIT(i);
+		}
+	}
+
+	return mask;
+}
+
+#define NTHREADS   CONFIG_MP_MAX_NUM_CPUS
+#define STACK_SIZE 1024
+#define PRIORITY   7
+
+static struct k_sem sems[NTHREADS];
+static struct k_sem done[NTHREADS];
+static struct k_thread threads[NTHREADS];
+K_THREAD_STACK_ARRAY_DEFINE(tstack, NTHREADS, STACK_SIZE);
+
+static void add(volatile uint32_t *x, const uint32_t val)
+{
+	*x += val;
+}
+
+/* This NOP macro generates at least 2 instructions. Notably if only nop.n is
+ * used the compiler will remove duplicate nop.n instructions so a volatile add
+ * is done to ensure this does not occur.
+ */
+#define NOP(i, val, ...)                                                                           \
+	add(&val, i);                                                                              \
+	__asm__ volatile("nop.n")
+
+static bool cpu_active[NTHREADS];
+
+static void do_work(void)
+{
+	volatile uint32_t v = 0;
+
+	/* These LISTIFY macros will ensure some number of actual cpu operations
+	 * will occur to simulate real work being done.
+	 * Can be copied/pasted a few times as needed to generate more op codes.
+	 */
+	LISTIFY(4096, NOP, (;), v);
+	LISTIFY(4096, NOP, (;), v);
+}
+
+void coreN_loop(void *arg0, void *arg1, void *arg2)
+{
+	uint32_t cnt = 0;
+	uint32_t cpu = arch_curr_cpu()->id;
+
+	while (true) {
+		printk("core %d waiting on semaphore\n", cpu);
+		k_sem_take(&sems[cpu],
+			   K_FOREVER); /* put core X to sleep until semaphore is given */
+		printk("core %d doing work\n", cpu);
+		do_work();
+		printk("core %d done\n", cpu);
+		k_sem_give(&done[cpu]);
+	}
+}
+
+const char *names[4] = {
+	"core0_loop",
+	"core1_loop",
+	"core2_loop",
+	"core3_loop",
+};
```

Comment:
This should be based on `NTHREADS`

Reply:
Yes, some listify/stringify stuff should make this array precisely sized with all the right strings, will fix